### ISSUE TITLE
[Proposal] Merge Mocktail into Mockito

### ISSUE THREAD

**Issue by felangel** [dart-lang/mockito#347]
Hello 👋 

I've been working with various members of the community for a while to brainstorm ways of maintaining the previous mockito API while still being null safe so that developers can migrate their apps more seamlessly without needing to use code generation or write lots of manual mocks (#286, #302).

The experimentation was all done under [package:mocktail](https://pub.dev/packages/mocktail).

The most recent [prerelease of mocktail](https://github.com/felangel/mocktail/pull/22) is very close to mockito and only has a few small breaking changes syntactically.

Previously in mockito
```dart
when(cat.sound()).thenReturn('meow');
verify(cat.sound()).called(1);
```

Currently in mocktail:

```dart
when(() => cat.sound()).thenReturn('meow');
verify(() => cat.sound()).called(1);
```

The most tricky aspect that we encountered so far is supporting the various arg matchers.

Previously in mockito:

```dart
when(calculator.sum(any, any)).thenReturn(42);
```

Currently in mocktail:

```dart
registerFallbackValue<int>(0);
when(() => calculator.sum(any(), any()).thenReturn(42);
```

**We're still working through the arg matchers and are open to any suggestions/feedback.

This allows mocktail to work around the type errors without the need for code generation (massive shoutout to @rrousselGit for the suggestion to use this approach).

Would the team be open to merging `mocktail` into `mockito`? I think it would be great for the community and also make the migration to null safety more seamless.

Thanks! 🙏

**Comment by TimWhiting** [comment]
Also this looks like something that a dart fix could automate, or a [codemod](https://pub.dev/packages/codemod) could be created.

**Comment by orestesgaolin** [comment]
This looks fantastic. Are there any substantial differences in API area? 

Combining `mocktail` with `mockito` would probably mean another major release, am I right?

**Comment by felangel** [comment]
> This looks fantastic. Are there any substantial differences in API area?
> 
> Combining `mocktail` with `mockito` would probably mean another major release, am I right?

Thanks! The main differences are instead of using the mock instance and invoking the method/getter/setter directly you'd need to do the invocation in a closure. Aside from that and the `registerFallbackValue` for the `any` and `captureAny` arg matchers, the API is the same but it would unfortunately be a breaking change and require another major release.

**Comment by srawlins** [comment]
Very cool project. I like the `registerFallbackValue` API.

> We're still working through the arg matchers and are open to any suggestions/feedback.

Can you go into detail here? What currently does not work in mocktail?

> `when(() => cat.sound()).thenReturn('meow');`
> `verify(() => cat.sound()).called(1);`

These breaking API changes are going to be hard to land in mockito. `dart fix` may help, but we need an incremental migration path. We need a way to support this new `() => cat.sound()` API while not breaking any existing usage of mockito.

> `when(() => calculator.sum(any(), any()).thenReturn(42);`

This one as well. In migrating to this API, we need to support this new API while not breaking existing usage of mockito.

**Comment by felangel** [comment]
@srawlins thanks for the quick response!

> Can you go into detail here? What currently does not work in mocktail?

As far as I know everything works in the current prerelease of mocktail. The current mockito test suite is included along with additional unit tests.

> These breaking API changes are going to be hard to land in mockito. dart fix may help, but we need an incremental migration path. We need a way to support this new () => cat.sound() API while not breaking any existing usage of mockito.

We could theoretically support both APIs simultaneously but the downside would be developers would lose type safety because we'd have to make the signature of when, verify, etc. very loose.

Just for my understanding, what would be the downside of a new major version of mockito? I know it would mean back-to-back breaking changes with major version bumps but if we agree that the proposed API is the way to go I feel it's worth it. Thanks again!

**Comment by srawlins** [comment]
> We could theoretically support both APIs simultaneously but the downside would be developers would lose type safety because we'd have to make the signature of when, verify, etc. very loose.

Right, I'm not crazy about that. Something like `when2`, `any2`, _cringe_, etc would allow type safety and then just force this dance of renamings. Something like:

* 5.1.0: add `when2`; deprecate `when`.
* 6.0.0: remove `when`.
* 6.1.0: add back `when`, which is just an alias for `when2`; deprecated `when2`.
* 7.0.0: remove `when2`.

> Just for my understanding, what would be the downside of a new major version of mockito? I know it would mean back-to-back breaking changes with major version bumps but if we agree that the proposed API is the way to go I feel it's worth it. Thanks again!

There is nothing wrong with a new major version of mockito.

For large mono-repos (like Google's codebase), it may be infeasible to introduce a breaking change without temporary backward compatibility.

**Comment by TimWhiting** [comment]
What about something like:

Have two libraries within the package:

`mockito/mockito.dart` -- with the new apis
`mockito/mockito_old.dart` -- with the old

That way you aren't doing a rename dance other than just changing the import.

Then deprecate the old library, and migrate the code to temporarily use `mockito_old.dart` until you have time to migrate.

Once code has been given time to migrate, do a major version bump and get rid of the old apis. (Potentially including a dart fix).

**Comment by felangel** [comment]
@TimWhiting I was just typing up the exact same suggestion 😄

**Comment by jtmcdole** [comment]
A seasoned engineer once told me, "avoid naming anything `new` or `old` because it will quickly be outdated".  Can we @deprecate the top of a library (mockito.dart) and name the new one mockito_nullsafety.dart?

**Comment by kuhnroyal** [comment]
Can we get some official guidance on this. What are the plans of the Dart team?

With all the good things that come with mull-safety and NNBD, the problems around mockito usage/generation are one of the biggest pain points. Big projects can't properly migrate because the tests just require an insane amount of work to be migrated with no easy path.

I have used mocktail in a couple smaller projects and while there is also work involved, it is in most cases pretty straight forward.
The package is gaining popularity pretty fast and it would just be great to know the feelings/expectations/directions of the Dart team.

**Comment by creativecreatorormaybenot** [comment]
To be frank @kuhnroyal, the migration to the null-safe version of `mockito` seems straight-forward as well.

For the majority of your tests (if not all of them), you can simply replace your `extends Mock` classes with one entry in the `@GenerateMocks` annotation. If you expected `null` on some of the calls before, you simply configure it with a `MockSpec`.  
I doubt that the migration to `mocktail` is easier than that.

**Comment by andrzejchm** [comment]
sorry @creativecreatorormaybenot , but every time I can choose between code generation and NO code generation, then the later wins by a mile. Currently, mocktail brings more value and convenience to the developer than the current version of mockito.

I've tried to use mockito without code generation with a null-safe code and its a nightmare. Code generation is tremendously slow in big projects unfortunately and requires you to run the `build` or `watch` command and remember about doing so

**Comment by jtmcdole** [comment]
I've started converting tests over to mockito's `GenerateMocks` and I've not found it complex. The most complex use case (a mock that offered to create stream controllers and set them up) required using the customMocks + `MockSpec` with a basename that we then extended as the "real" mock. We have several shared mocks in our build system; and for those I've found adding `@GenerateMocks` to the shared libraries and exporting the mocks worked great, rather than at the head of each test file.

**Comment by felangel** [comment]
> I've started converting tests over to mockito's `GenerateMocks` and I've not found it complex. The most complex use case (a mock that offered to create stream controllers and set them up) required using the customMocks + `MockSpec` with a basename that we then extended as the "real" mock. We have several shared mocks in our build system; and for those I've found adding `@GenerateMocks` to the shared libraries and exporting the mocks worked great, rather than at the head of each test file.

In my opinion the syntactical changes are not the main issue. The fact that code generation is involved when writing tests doesn’t scale well in my opinion. In extremely large codebases with thousands of tests the build_runner process alone will take several minutes and new mocks will likely need to be generated frequently.

**Comment by creativecreatorormaybenot** [comment]
Just want to point out that I think it is hilarious that the people who agree with @andrzejchm are 👎 voting my comment.  

I never said I think code gen is preferrable, but why read my comment if you can just downvote it. I have been an advocate of `mocktail` even before it was released, but this does not change the fact that migrating from `mockito: 4.0.0` to `mocktail` is more complex than migrating to `mockito: 5.0.0`. Thanks.

**Comment by creativecreatorormaybenot** [comment]
Just want to point out that I think it is hilarious that the people who agree with @andrzejchm are 👎 voting my comment.  

I never said I think code gen is preferrable, but why read my comment if you can just downvote it. I have been an advocate of `mocktail` even before it was released (and you can see my name if you hover over the upvotes on this issue as I upvoted as one of the first people), but this does not change the fact that migrating from `mockito: 4.0.0` to `mocktail` is more complex than migrating to `mockito: 5.0.0`. Thanks.

**Comment by creativecreatorormaybenot** [comment]
Just want to point out that I think it is hilarious that the people who agree with @andrzejchm are 👎 voting my comment.  

I never said I think code gen is preferrable, but **why read my comment if you can just downvote it**.  
I have been an advocate of `mocktail` even before it was released (and you can see my name if you hover over the upvotes on this issue as I upvoted as one of the first people), but this does not change the fact that migrating from `mockito: 4.0.0` to `mocktail` is more complex than migrating to `mockito: 5.0.0`. Thanks.

**Comment by felangel** [comment]
> Just want to point out that I think it is hilarious that the people who agree with @andrzejchm are 👎 voting my comment.
> 
> I never said I think code gen is preferrable, but why read my comment if you can just downvote it. I have been an advocate of `mocktail` even before it was released, but this does not change the fact that migrating from `mockito: 4.0.0` to `mocktail` is more complex than migrating to `mockito: 5.0.0`. Thanks.

Migrating from mockito to mocktail can be automated with a codemod and once the migration is done the developer experience is almost identical to the way it was before. Sorry for any misunderstanding!

The goal of this issue was just to understand how the Dart team feels about this approach rather than relying on code gen. If the decision is to keep the current mockito api I can close this issue.

**Comment by rrousselGit** [comment]
On a different note, with static-metaprogramming, the slowness of code-generation will solve itself and Mockito will likely take the lead again.


So it may not be worth breaking users now when we could not break them in a few months.

**Comment by creativecreatorormaybenot** [comment]
@felangel I know and I agree - that was the point of my message.

**Comment by jtmcdole** [comment]
> In my opinion the syntactical changes are not the main issue. The fact that code generation is involved when writing tests doesn’t scale well in my opinion. In extremely large codebases with thousands of tests the build_runner process alone will take several minutes and new mocks will likely need to be generated frequently.

We have tons of tests that run with multiple developers; granted Google infrastructure is bonkers and really spoils you.

**Comment by creativecreatorormaybenot** [comment]
@jtmcdole But this is a good point - how do you run these test? Can you share anything about that?

That is:
1. Do you run them locally at all?
2. How does the "bonkers Google infrastructure" handle the CI? 😄

**Comment by cedvdb** [comment]
> On a different note, with static-metaprogramming, the slowness of code-generation will solve itself and Mockito will likely take the lead again.
> 
> So it may not be worth breaking users now when we could not break them in a few months.

That's a very optimistic deadline. I'm glad to hear that. I was envisioning next year.

**Comment by felangel** [comment]
@srawlins @jtmcdole do you have any sense of what the status of this issue is? If it's unlikely for `mocktail` to be merged into `mockito` it would be really helpful to know and this issue can be closed. Thanks!

**Comment by srawlins** [comment]
Apologies for the extended delay in keeping up with this thread.

I think the two big features from Mocktail that are desirable in mockito are:

1. stub and verify a method call, for a method with a non-nullable return type, (`when(() => cat.sound()).thenReturn('meow')`) without code generation,
2. stub and verify a method call with argument matchers via an API which stores an argument matcher and passes a real value of a non-nullable type to the method (`registerFallbackValue`).

I'll address them in reverse order.

### 2. non-nullable parameter types via `registerFallbackValue`

This feature seems to be more ergonomic than I originally thought it would be. In many cases, mockito already works in a "global" sense, and the global fallback value list seems very simple. The implementation of fetching a fallback value is _O(n)_ with the number of fallback values which I don't think is too bad, because (a) the list starts out with ~5 values and it's hard to imagine it growing past 100 or so in a typical test run, or even 500 in an extreme case... and (b) its just used in tests, not production code.

I wonder though, how do argument matchers work in concert with real values? I can't find a test that quite fits what I mean, which is something like `foo(int a, int b) {} ... when(() => foo(any(), 42))`. Firstly, can mocktail handle this (I wouldn't expect it to; mockito cannot handle `when(foo(any, null))`)? And if it cannot, can it detect the situation and fail with a helpful error like "I got one argument matcher but two arguments which look like argument matchers; perhaps you passed a 42 value?" 42 maybe is not a common literal value, but what about `foo(List a, List b) {} ... when(() => foo(any(), []))` or worse yet, bool values?

Why is `registerFallbackValue` generic?

In general, the list of fallback values is just sort of this pile of sentinels, and it never really matters which one you pick, right? Meaning that, if the user has registered `<int>[]`, `<num>[]`, `<int>{}`, `<num>{}`, and `<String>[]`, and then passes `any()` for an `Iterable<Object?>` parameter, it doesn't matter which fallback is retrieved, as far as matching the argument matcher with the parameter goes, as long as _some_ value is retrieved.

### 1. non-nullable return type

This API change is breaking (on which most of this thread has been centered); I really don't think we can land it as is, as a replacement for `when`. The incremental migration would be a lot of work (for Google code) without benefit (as it all uses GenerateMocks). But we may have some options:

1. Literally duplicate the stubbing and verification APIs (`when`, `verify`, `verifyInOrder`, `verifyNever`, I think is the complete list), with APIs that take closures and execute them, as mocktail does. Something like `whenCalled`, `whenCalling`, `verifyCalled`, `verifyCalledInOrder`, or better names :). The price we pay here is maintenance of more APIs. I think it would mean nearly duplicating the tests. It also adds complexity for users, but we can alleviate some of that with good documentation, and possibly by providing this set of APIs in a different top-level library.

2. OK I got halfway through two other ideas and they got nowhere. :(

**Comment by rrousselGit** [comment]
> Why is registerFallbackValue generic?

This makes obtaining the fallback value O(1) instead of O(N)
Internally fallback are stored in a `Map<Type, Object?>`

Although your idea of "try everything in the pile and see what sticks" is interesting.
But I wonder if this could cause issues for generic classes, since while `List<int>` can be assigned to `List<Object>`, they don't really behave the same way.

**Comment by srawlins** [comment]
> Internally fallback are stored in a `Map<Type, Object?>`

Er, where is that? I'm seeing a List and `firstWhereOrNull`.

https://github.com/felangel/mocktail/blob/main/packages/mocktail/lib/src/_register_matcher.dart#L34..L47

**Comment by rrousselGit** [comment]
Ah, looks like it has changed since the last time I touched the code.
My bad then

**Comment by andreiborisov** [comment]
👋 We're currently deciding on the mock library for our current project.

It would be helpful to get a sense of the direction you're going in with this, so people can choose an API that would require less migration in the future.

**Comment by jeroen-meijer** [comment]
Rereading this a while later, looks like the merge won't happen anytime soon, right?
If so, we can close this if you ask me.

**Comment by srawlins** [comment]
It won't happen anytime soon, no.
